pkg/kubelet/cm/cpumanager/cpu_manager.go

```diff
@@ -206,72 +227,123 @@ func (m *manager) State() state.Reader {
 	return m.state
 }
 
-type reconciledContainer struct {
-	podName       string
-	containerName string
-	containerID   string
+func (m *manager) reconcile() {
+	// forever
+	for {
+		select {
+		case <-m.reconcileAll:
+			glog.V(5).Info("[cpumanager] reconciling all active containers")
+			m.resetContainersToReconcile()
+			m.containersToReconcile = m.doReconcile()
+		case <-m.reconcileFailed:
+			glog.V(5).Info("[cpumanager] reconciling containers that previously failed to reconcile")
+			m.containersToReconcile = m.doReconcile()
+		}
+	}
 }
 
-func (m *manager) reconcileState() (success []reconciledContainer, failure []reconciledContainer) {
-	success = []reconciledContainer{}
-	failure = []reconciledContainer{}
+// Implements ReconcileFunc; this function is passed to the Policy.
+func (m *manager) reconcileFunc() {
+	m.reconcileAll <- struct{}{}
+}
 
+func (m *manager) resetContainersToReconcile() {
+	result := make(map[*v1.Pod][]v1.Container)
 	for _, pod := range m.activePods() {
 		allContainers := pod.Spec.InitContainers
 		allContainers = append(allContainers, pod.Spec.Containers...)
-		for _, container := range allContainers {
-			status, ok := m.podStatusProvider.GetPodStatus(pod.UID)
-			if !ok {
-				glog.Warningf("[cpumanager] reconcileState: skipping pod; status not found (pod: %s, container: %s)", pod.Name, container.Name)
-				failure = append(failure, reconciledContainer{pod.Name, container.Name, ""})
-				break
+		result[pod] = allContainers
+	}
+	m.containersToReconcile = result
+}
+
+// This logic is thread-safe by virtue of only being called from
+// reconcile(), which is itself executed in only one goroutine.
+func (m *manager) doReconcile() (failed map[*v1.Pod][]v1.Container) {
+	failed = make(map[*v1.Pod][]v1.Container)
+
+	// Adds the supplied pod and container to the list of failed
+	// reconciliations.
+	addFailed := func(pod *v1.Pod, container v1.Container) {
+		if _, ok := failed[pod]; !ok {
+			failed[pod] = make([]v1.Container, 1)
+		}
+		podContainers := failed[pod]
+		failed[pod] = append(podContainers, container)
+	}
+
+	// Returns true if the supplied pod exists in the list of active pods.
+	podExists := func(activePods []*v1.Pod, pod *v1.Pod) bool {
+		for _, activePod := range activePods {
+			if activePod.UID == pod.UID {
```

Comment:
Looks like `activePods` could be a map...

Reply:
It comes from elsewhere in the kubelet, so this isn't feasible. The default cap on number of pods per node was 110 last I looked; iterating such a small list is not worth optimizing IMO.